openhands/sdk/llm/router/impl/multimodal.py

```diff
@@ -0,0 +1,84 @@
+from typing import ClassVar
+
+from pydantic import model_validator
+
+from openhands.sdk.llm.message import Message
+from openhands.sdk.llm.router.base import RouterLLM
+from openhands.sdk.logger import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class MultimodalRouter(RouterLLM):
+    """
+    A RouterLLM implementation that routes requests based on multimodal content
+    (e.g., images) and token limits. If any message contains multimodal content
+    or if the token limit of the secondary model is exceeded, it routes to the
+    primary model. Otherwise, it routes to the secondary model.
+
+    Note: The primary model is expected to support multimodal content, while
+    the secondary model is typically a text-only model with a lower context window.
+    """
+
+    router_name: str = "multimodal_router"
+
+    PRIMARY_MODEL_KEY: ClassVar[str] = "primary"
+    SECONDARY_MODEL_KEY: ClassVar[str] = "secondary"
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self._max_token_exceeded = False
+
+    def _select_llm(self, messages: list[Message]) -> str:
+        """Select LLM based on multimodal content and token limits."""
+        route_to_primary = False
+
+        # Check for multimodal content in messages
+        for message in messages:
+            if message.contains_image:
+                logger.info(
+                    "Multimodal content detected in messages. "
+                    "Routing to the primary model."
+                )
+                route_to_primary = True
+
+        if not route_to_primary and self._max_token_exceeded:
+            route_to_primary = True
+
+        # Check if `messages` exceeds context window of the secondary model
+        # Assuming the secondary model has a lower context window limit
+        # compared to the primary model
+        secondary_llm = self.llms_for_routing.get(self.SECONDARY_MODEL_KEY)
+        if secondary_llm and (
+            secondary_llm.max_input_tokens
+            and secondary_llm.get_token_count(messages) > secondary_llm.max_input_tokens
+        ):
+            logger.warning(
+                f"Messages having {secondary_llm.get_token_count(messages)} tokens, exceeded secondary model's max input tokens ({secondary_llm.max_input_tokens} tokens). "  # noqa: E501
+                "Routing to the primary model."
+            )
+            self._max_token_exceeded = True
```

Comment:
we shouldn't maintain it as a state here-- it is possible that a condenser kicks in, the input token will drop in the next request -- in that case, this won't be updated & request will still be routed to primary even though it can be route to secondary, right?